### PR TITLE
feat(Isolation): Added isolation support

### PR DESCRIPTION
**Pull request by zhenlei520**
### Chore:
* IntegrationEvents.Dapr: Adjust UoW enable check prompt (#28)
* UoW.EF: remove invalid references (#28)

### Docs
* Isolation: Add isolation Readme doc (#28)

### Feature:
* Isolation: Added isolation support (#28)
* UoW.EF: CreateDbContext supports lazy loading (#28)

### Test
* Isolation: add Isolation Unit tests (#28)

### CODE REVIEW

File: src/Isolation/Masa.Contrib.Isolation.MultiTenancy/README.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
+[中](README.zh-CN.md) | EN
+
+## Masa.Contrib.Isolation.MultiTenancy
```
Comment (doddgu) [inline src/Isolation/Masa.Contrib.Isolation.MultiTenancy/README.md:3]: MultiTenant? MultiTenancy?

File: src/Isolation/Masa.Contrib.Isolation.MultiTenancy/README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        "TenantId": "00000000-0000-0000-0000-000000000003",
+        "ConnectionString": "server=localhost,1672;uid=sa;pwd=P@ssw0rd;database=identity;"
+      }
+    ]
+  }
+}
+```
+
+* 1.1 When the current tenant is 00000000-0000-0000-0000-000000000002: database address: server=localhost,1674;uid=sa;pwd=P@ssw0rd;database=identity;
+* 1.2 When the current tenant is 00000000-0000-0000-0000-000000000003: database address: server=localhost,1672;uid=sa;pwd=P@ssw0rd;database=identity;
```
Comment (doddgu) [inline src/Isolation/Masa.Contrib.Isolation.MultiTenancy/README.md:33]: 1.3 others?

File: src/Isolation/Masa.Contrib.Isolation.MultiTenancy/TenancySaveChangesFilter.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+﻿namespace Masa.Contrib.Isolation.MultiTenancy;
+
+public class TenancySaveChangesFilter<TKey> : ISaveChangesFilter where TKey : IComparable
```
Comment (doddgu) [inline src/Isolation/Masa.Contrib.Isolation.MultiTenancy/TenancySaveChangesFilter.cs:3]: Tenant? Tenancy?

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    private string GetMessage()
+    {
+        StringBuilder stringBuilder = new StringBuilder();
+        if (_environmentContext != null)
+            stringBuilder.Append($"Environment: [{_environmentContext.CurrentEnvironment}], ");
+        if (_tenantContext != null)
+            stringBuilder.Append($"Tenant: [{_tenantContext.CurrentTenant?.Id ?? ""}]");
+        var message = stringBuilder.ToString();
+        return message.Substring(0, message.Length - 1);
```
Comment (doddgu) [inline src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs:86]: "Tenant: [xxx]".Length - 1? And both null, Length - 1 = -1?

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/Internal/TypeCommon.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Masa.Contrib.Isolation.UoW.EF.Internal;
+
+internal static class TypeCommon
```
Comment (doddgu) [inline src/Isolation/Masa.Contrib.Isolation.UoW.EF/Internal/TypeCommon.cs:3]: Common is a bad name

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/Parser/MultiTenancy/HeaderTenantParserProvider.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
+﻿namespace Masa.Contrib.Isolation.UoW.EF.Parser.MultiTenancy;
+
+public class HeaderTenantParserProvider : ITenantParserProvider
+{
+    public string Name => "Header";
```
Comment (doddgu) [inline src/Isolation/Masa.Contrib.Isolation.UoW.EF/Parser/MultiTenancy/HeaderTenantParserProvider.cs:5]: How to use for each Parser?

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/README.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+> When the DbContext is not operated in the EventHandler, it needs to be used when the DbContext is directly operated in the Controller or Minimal, otherwise the environment and tenant acquisition will fail, resulting in the database address being selected as the default DefaultConnection address
+
+4. The class corresponding to the isolated table needs to implement IMultiTenant or IEnvironment
+
+Tenant isolation implements IMultiTenant, and environment isolation implements IEnvironment
+
+##### Summarize
+* When are tenants and environments resolved?
+    * Manipulating DbContext, Repository, etc. in EventHandler can be used directly without adding. The environment and tenant information will be parsed when the Event is published. If the environment or tenant has been assigned, the parsing will be skipped
```
Comment (doddgu) [inline src/Isolation/Masa.Contrib.Isolation.UoW.EF/README.md:68]: Without adding what?

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs (hunk 4 above)
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs:86]: stringBuilder cannot be null, at least one of _environmentContext and _tenantContext exists, which is restricted in UseIsolationUoW

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/Internal/TypeCommon.cs (hunk 5 above)
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/Internal/TypeCommon.cs:3]: processed

File: src/Isolation/Masa.Contrib.Isolation.Environment/README.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
+[中](README.zh-CN.md) | EN
+
+## Masa.Contrib.Isolation.Environment
+
+Example：
+
+```C#
+Install-Package Masa.Contrib.Isolation.UoW.EF
+Install-Package Masa.Contrib.Isolation.Environment // Environmental isolation Quote on demand
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.Environment/README.md:9]: processed

File: src/Isolation/Masa.Contrib.Isolation.Environment/README.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+```` C#
+builder.Services.AddEventBus(eventBusBuilder =>
+{
+    eventBusBuilder.UseIsolationUoW<CustomDbContext>(
+        dbOptions => dbOptions.UseSqlServer(),
+        isolationBuilder => isolationBuilder.UseEnvironment());// Use environment isolation
+});
+````
+
+3. DbContext needs to inherit IsolationDbContext
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.Environment/README.md:45]: The sample code of the documentation is wrong, MasaDbContext does not inherit IsolationDbContext

File: src/Isolation/Masa.Contrib.Isolation.Environment/README.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+{
+    public CustomDbContext(MasaDbContextOptions<CustomDbContext> options) : base(options)
+    {
+    }
+}
+````
+
+4. The class corresponding to the isolated table needs to implement IEnvironment
+
+You can also choose not to implement IMultiTenant when using physical isolation
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.Environment/README.md:58]: The code of the document is wrong, here is the IMultiEnvironment

File: src/Isolation/Masa.Contrib.Isolation.Environment/README.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+```` C#
+public class CustomDbContext : MasaDbContext
+{
+    public CustomDbContext(MasaDbContextOptions<CustomDbContext> options) : base(options)
+    {
+    }
+}
+````
+
+4. The class corresponding to the isolated table needs to implement IEnvironment
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.Environment/README.md:56]: processed

File: src/Isolation/Masa.Contrib.Isolation.MultiTenancy/ConvertProvider.cs [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    public object ChangeType(string value, Type conversionType)
+    {
+        object result;
+        if (conversionType == typeof(Guid))
+        {
+            result = Guid.Parse(value);
+        }
+        else
+        {
+            result= Convert.ChangeType(value, conversionType);
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.MultiTenancy/ConvertProvider.cs:14]: processed

File: src/Isolation/Masa.Contrib.Isolation.MultiTenancy/README.md (hunk 1 above)
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.MultiTenancy/README.md:3]: Uniformly modified to MultiTenant, processed

File: src/Isolation/Masa.Contrib.Isolation.MultiTenancy/README.md (hunk 2 above)
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.MultiTenancy/README.md:33]: processed

File: src/Isolation/Masa.Contrib.Isolation.MultiTenancy/TenancySaveChangesFilter.cs (hunk 3 above)
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.MultiTenancy/TenancySaveChangesFilter.cs:3]: processed

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/README.md (hunk 7 above)
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/README.md]: processed

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/Parser/MultiTenancy/HeaderTenantParserProvider.cs (hunk 6 above)
Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/Parser/MultiTenancy/HeaderTenantParserProvider.cs]: Need to explain how each parser is used here?

File: src/Data/Masa.Contrib.Data.UoW.EF/UnitOfWorkManager.cs [hunk 13]
```diff
@@ -1,14 +1,22 @@
...
-public class UnitOfWorkManager : IUnitOfWorkManager
+public class UnitOfWorkManager<TDbContext> : IUnitOfWorkManager where TDbContext : MasaDbContext
 {
     private readonly IServiceProvider _serviceProvider;
 
     public UnitOfWorkManager(IServiceProvider serviceProvider) => _serviceProvider = serviceProvider;
 
-    public IUnitOfWork CreateDbContext()
+    /// <summary>
+    ///
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Data/Masa.Contrib.Data.UoW.EF/UnitOfWorkManager.cs]: processed

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs (hunk 4 above)
Comment (doddgu) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs]: "Tenant: [xxx]".Length - 1?

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/Parser/MultiTenancy/HeaderTenantParserProvider.cs (hunk 6 above)
Comment (doddgu) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/Parser/MultiTenancy/HeaderTenantParserProvider.cs]: doc

Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/Parser/MultiTenancy/HeaderTenantParserProvider.cs]: ok, the parser here is very simple, so I did not specifically explain how to use the parser, I will add it later

File: src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs (hunk 4 above)
Comment (doddgu) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs]: "Tenant: [xxx],".Length - 1?  https://makolyte.com/csharp-join-strings-with-a-separator-ignoring-nulls-and-empty-strings/

Comment (doddgu) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs]: Too ugly, https://makolyte.com/csharp-join-strings-with-a-separator-ignoring-nulls-and-empty-strings/

Comment (zhenlei520) [reply src/Isolation/Masa.Contrib.Isolation.UoW.EF/DefaultConnectionStringProvider.cs]: processed

**Review by doddgu — APPROVED** [review]
LGTM

**Pull request merged**